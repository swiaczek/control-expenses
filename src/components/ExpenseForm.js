import React, { Component } from "react";
import moment from "moment";
import "react-dates/initialize";
import "react-dates/lib/css/_datepicker.css";
import { SingleDatePicker } from "react-dates";

class ExpenseForm extends Component {
  constructor(props) {
    super(props);

    this.state = {
      description: props.expense ? props.expense.description : "",
      amount: props.expense ? (props.expense.amount / 100).toString() : "",
      note: props.expense ? props.expense.note : "",
      createdAt: props.expense ? moment(props.expense.createdAt) : moment(),
      error: false,
      focusCalendar: false,
    };
  }
  handleDescriptionChange = (e) => {
    const description = e.target.value;
    this.setState(() => ({ description }));
  };

  handleAmountChange = (e) => {
    const amount = e.target.value;
    if (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
      this.setState(() => ({ amount }));
    }
  };

  handleNoteChange = (e) => {
    const note = e.target.value;
    this.setState(() => ({ note }));
  };

  onDateChange = (createdAt) => {
    if (createdAt) {
      this.setState(() => ({ createdAt }));
    }
  };

  onFocusChange = ({ focused }) => {
    this.setState({ focusCalendar: focused });
  };

  handleSubmitForm = (e) => {
    e.preventDefault();
    if (!this.state.description || !this.state.amount) {
      this.setState(() => ({ error: true }));
    } else {
      this.setState(() => ({ error: false }));
      this.props.onSubmit({
        description: this.state.description,
        amount: parseFloat(this.state.amount, 10) * 100,
        createdAt: this.state.createdAt.valueOf(),
        note: this.state.note,
      });
    }
  };

  render() {
    return (
      <div>
        {this.state.error ? (
          <p style={{ color: "red" }}>Please provide description and amount</p>
        ) : null}
        <form onSubmit={this.handleSubmitForm}>
          <input
            type="text"
            placeholder="Description"
            autoFocus
            value={this.state.description}
            onChange={this.handleDescriptionChange}
          />
          <input
            type="text"
            placeholder="Amount"
            value={this.state.amount}
            onChange={this.handleAmountChange}
          />
          <SingleDatePicker
            date={this.state.createdAt}
            onDateChange={this.onDateChange}
            focused={this.state.focusCalendar}
            onFocusChange={this.onFocusChange}
            id="date_Id"
            numberOfMonths={1}
            isOutsideRange={() => false}
          />
          <textarea
            placeholder="Add a note for your expense (optional)"
            value={this.state.note}
            onChange={this.handleNoteChange}
          ></textarea>
          <button>Add Expense</button>
        </form>
      </div>
    );
  }
}

export default ExpenseForm;
