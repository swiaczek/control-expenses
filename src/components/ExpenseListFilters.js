import React, { Component } from "react";
import { connect } from "react-redux";
import {
  setTextFilter,
  sortByAmount,
  sortByDate,
  setStartDate,
  setEndDate,
} from "../actions/filters";
import "react-dates/initialize";
import "react-dates/lib/css/_datepicker.css";
import { DateRangePicker } from "react-dates";

export class ExpenseListFilters extends Component {
  state = {
    focusCalendar: null,
  };

  onDatesChange = ({ startDate, endDate }) => {
    this.props.setStartDate(startDate);
    this.props.setEndDate(endDate);
  };

  onFocusChange = (focusCalendar) => {
    this.setState({ focusCalendar });
  };

  onSortChange = (e) => {
    if (e.target.value === "date") {
      return this.props.sortByDate();
    } else if (e.target.value === "amount") {
      return this.props.sortByAmount();
    }
  };

  onTextChange = (e) => {
    this.props.setTextFilter(e.target.value);
  };
  render() {
    return (
      <div className="content-container">
        <div className="input-box">
          <div className="input-box__item">
            <input
              className="text-input"
              type="text"
              placeholder="Search expense..."
              value={this.props.filters.text}
              onChange={this.onTextChange}
            />
          </div>
          <div className="input-box__item">
            <select
              className="select"
              value={this.props.filters.sortBy}
              onChange={this.onSortChange}
            >
              <option value="date">Date</option>
              <option value="amount">Amount</option>
            </select>
          </div>
          <div className="input-box__item">
            <DateRangePicker
              startDate={this.props.filters.startDate}
              startDateId="start_Id"
              endDate={this.props.filters.endDate}
              endDateId="end_Id"
              onDatesChange={this.onDatesChange}
              focusedInput={this.state.focusCalendar}
              onFocusChange={this.onFocusChange}
              numberOfMonths={1}
              isOutsideRange={() => false}
              showClearDates={true}
            />
          </div>
        </div>
      </div>
    );
  }
}

const mapStateToProps = (state) => {
  return {
    filters: state.filters,
  };
};

const mapDispatchToProps = (dispatch) => ({
  setTextFilter: (text) => dispatch(setTextFilter(text)),
  sortByAmount: () => dispatch(sortByAmount()),
  sortByDate: () => dispatch(sortByDate()),
  setStartDate: (startDate) => dispatch(setStartDate(startDate)),
  setEndDate: (endDate) => dispatch(setEndDate(endDate)),
});

export default connect(mapStateToProps, mapDispatchToProps)(ExpenseListFilters);
