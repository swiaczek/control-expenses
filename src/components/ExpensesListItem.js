import React from "react";
import { Link } from "react-router-dom";
import moment from "moment";
import numeral from "numeral";

const ExpensesListItem = ({ id, description, amount, createdAt }) => {
  return (
    <div>
      <Link to={`/edit/${id}`}>
        <h3>{description}</h3>
      </Link>
      <h5>
        {numeral(amount / 100).format("0,0.0")}â‚¬ -{" "}
        {moment(createdAt).format("Do MMM, YYYY")}
      </h5>
    </div>
  );
};

export default ExpensesListItem;
