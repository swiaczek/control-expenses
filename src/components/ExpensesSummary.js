import React from "react";
import { connect } from "react-redux";
import selectTotalExpenses from "../selectors/expenses-total";
import selectExpenses from "../selectors/expenses";
import numeral from "numeral";

export const ExpensesSummary = ({ totalCount, totalExpenses }) => {
  const expenseWord = totalCount === 1 ? "expense" : "expenses";
  const formattedTotalExpenses = numeral(totalExpenses / 100).format("0,0.0");
  return (
    <div>
      <h3>
        Viewing {totalCount} {expenseWord} totalling {formattedTotalExpenses}â‚¬
      </h3>
    </div>
  );
};

const mapStateToProps = (state) => {
  const visibleExpenses = selectExpenses(state.expenses, state.filters);
  return {
    totalCount: visibleExpenses.length,
    totalExpenses: selectTotalExpenses(visibleExpenses),
  };
};

export default connect(mapStateToProps)(ExpensesSummary);
