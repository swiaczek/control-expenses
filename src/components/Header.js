import React from "react";
import { NavLink } from "react-router-dom";
import { connect } from "react-redux";
import { startLogout } from "../actions/auth";

export const Header = ({ startLogout }) => {
  return (
    <header>
      <h1>Control Expenses</h1>
      <NavLink to="/dashboard" activeClassName="is-active">
        Dashboards
      </NavLink>
      <NavLink to="/add" activeClassName="is-active">
        Create Expenses
      </NavLink>
      <NavLink to="/help" activeClassName="is-active">
        Help
      </NavLink>
      <button onClick={startLogout}>Logout</button>
    </header>
  );
};

const mapDispatchToProps = (dispatch) => ({
  startLogout: () => dispatch(startLogout()),
});

export default connect(undefined, mapDispatchToProps)(Header);
