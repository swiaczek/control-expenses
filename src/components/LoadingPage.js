import React from "react";
import loader from "../images/loader.gif";

const LoadingPage = () => {
  return (
    <div className="loader">
      <img src={loader} alt="loader" />
    </div>
  );
};

export default LoadingPage;
