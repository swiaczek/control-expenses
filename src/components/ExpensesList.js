import React from "react";
import { connect } from "react-redux";
import selectExpenses from "../selectors/expenses";
import ExpensesListItem from "./ExpensesListItem";

export const ExpensesList = (props) => {
  return (
    <div className="content-container">
      <div className="list-header">
        <div className="visable-for-desktop">Expense</div>
        <div className="visable-for-desktop">Amount</div>
        <div className="visable-for-mobile">Expenses</div>
      </div>
      {props.expenses.length ? (
        props.expenses.map((expense) => {
          return <ExpensesListItem key={expense.id} {...expense} />;
        })
      ) : (
        <span className="list-item list-item--message">No expenses</span>
      )}
    </div>
  );
};

const mapStateToProps = (state) => {
  return {
    expenses: selectExpenses(state.expenses, state.filters),
  };
};

export default connect(mapStateToProps)(ExpensesList);
