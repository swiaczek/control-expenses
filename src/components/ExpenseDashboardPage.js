import React from "react";
import ExpensesList from "./ExpensesList";
import ExpensesListFilters from "./ExpenseListFilters";
import ExpensesSummary from "./ExpensesSummary";

const ExpenseDashboardPage = () => {
  return (
    <div>
      <ExpensesSummary />
      <ExpensesListFilters />
      <ExpensesList />
    </div>
  );
};

export default ExpenseDashboardPage;
