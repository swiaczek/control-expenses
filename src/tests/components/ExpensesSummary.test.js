import React from "react";
import { shallow } from "enzyme";
import { ExpensesSummary } from "../../components/ExpensesSummary";

test("should render ExpensesSummary correctly with 1 expense", () => {
  const wrapper = shallow(
    <ExpensesSummary totalCount={1} totalExpenses={20000} />
  );
  expect(wrapper).toMatchSnapshot();
});

test("should render ExpensesSummary correctly with mulitplay expense", () => {
  const wrapper = shallow(
    <ExpensesSummary totalCount={20} totalExpenses={20000} />
  );
  expect(wrapper).toMatchSnapshot();
});
