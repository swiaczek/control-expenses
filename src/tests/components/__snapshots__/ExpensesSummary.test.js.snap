// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render ExpensesSummary correctly with 1 expense 1`] = `
<div>
  <h3>
    Viewing 
    1
     
    expense
     totalling 
    200.0
    €
  </h3>
</div>
`;

exports[`should render ExpensesSummary correctly with mulitplay expense 1`] = `
<div>
  <h3>
    Viewing 
    20
     
    expenses
     totalling 
    200.0
    €
  </h3>
</div>
`;
